test/utils/deployment.go

```diff
@@ -199,7 +199,15 @@ func WaitForDeploymentRevisionAndImage(c clientset.Interface, ns, deploymentName
 		return fmt.Errorf("deployment %q failed to create new replica set", deploymentName)
 	}
 	if err != nil {
-		return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deployment.Spec.Template.Spec.Containers[0].Image, newRS.Name, newRS.Annotations[deploymentutil.RevisionAnnotation], newRS.Spec.Template.Spec.Containers[0].Image, revision, image, err)
+		deploymentImage := ""
+		if deployment != nil && len(deployment.Spec.Template.Spec.Containers) > 0 {
+			deploymentImage = deployment.Spec.Template.Spec.Containers[0].Image
+		}
+		newRSImage := ""
+		if newRS != nil && len(newRS.Spec.Template.Spec.Containers) > 0 {
+			newRSImage = newRS.Spec.Template.Spec.Containers[0].Image
+		}
+		return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deploymentImage, newRS.Name, newRS.Annotations[deploymentutil.RevisionAnnotation], newRSImage, revision, image, err)
```

Comment:
the deployment == nil check should be before the err != nil

```
...
if newRS == nil {
	return fmt.Errorf("deployment %q failed to create new replica set", deploymentName)
}
if deployment == nil {
	return fmt.Errorf("could not get deployment %q", deploymentName)
}
if err != nil {
...
```

another improvement would be to expose the underlying error from the poll.

Reply:
newRS.Name is still accessed directly and could NPE

Reply:
fixed.

Reply:
are you sure? I still see it here:

return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deploymentImage, **newRS.Name**, newRS.Annotations[deploymentutil.RevisionAnnotation], newRSImage, revision, image, err)

Reply:
@liggitt we bail out in line 199 i think `newRS == nil`

Reply:
ah, got it